Problem with "using namespace std" in generated dictionary files

In the autogenerated dictionary sources I see

```
// Since CINT ignores the std namespace, we need to do so in this file.
namespace std {} using namespace std;
```

(this is from 6.14.08. 6.22.02 has a slightly different comment of `The generated code does not explicitly qualifies STL entities` (sic), but still does the `using...`)

This is placed very early on, even before the include of the header files passed as arguments.

Unfortunately, this causes problems, specifically when one of those header files pulls in Kokkos core headers, and there's a name clash with `rank` which is a free function in the kokkos namespace, and std::rank.

```
/opt/kokkos/cuda/3.2-shlib/include/Kokkos_View.hpp:1896:30: error: return type specified for deduction guide
 KOKKOS_INLINE_FUNCTION constexpr unsigned rank(const View<D, P...>& V) {
                              ^
/opt/kokkos/cuda/3.2-shlib/include/Kokkos_View.hpp:1895:42: error: decl-specifier in declaration of deduction guide
 template <typename D, class... P>
                                          ^       
/opt/kokkos/cuda/3.2-shlib/include/Kokkos_View.hpp:1895:32: error: decl-specifier in declaration of deduction guide
 template <typename D, class... P>
                                ^~       
/opt/kokkos/cuda/3.2-shlib/include/Kokkos_View.hpp:1896:30: error: 'signed' or 'unsigned' invalid for '__dguide_rank'
 KOKKOS_INLINE_FUNCTION constexpr unsigned rank(const View<D, P...>& V) {
                              ^
/opt/kokkos/cuda/3.2-shlib/include/Kokkos_View.hpp:1896:1: error: deduction guide for 'std::rank< <template-parameter-1-1> >' must have trailing return type
 KOKKOS_INLINE_FUNCTION constexpr unsigned rank(const View<D, P...>& V) {
 ^~~~
/bld4/opt/gcc/8.4.0/include/c++/8.4.0/type_traits:1251:8: note: 'template<class> struct std::rank' declared here
     struct rank
        ^~~~
```


Is there any way to make the `using namespace` appear AFTER the include of the user header files instead of before?

If I manually edit the generated file, and move the `using namespace` to after the includes,  it compiles fine.

FWIW, the actual compilation is done with nvcc 11, and c++17.

thanks, Charles.

--- a/core/dictgen/src/rootcling_impl.cxx
+++ b/core/dictgen/src/rootcling_impl.cxx
@@ -2830,7 +2830,7 @@ void CreateDictHeader(std::ostream &dictStream, const std::string &main_dictname
 
 void AddNamespaceSTDdeclaration(std::ostream &dictStream)
 {
-   dictStream  << "// The generated code does not explicitly qualifies STL entities\n"
+   dictStream  << "// The generated code does not explicitly qualify STL entities\n"
                << "namespace std {} using namespace std;\n\n";
 }
 
@@ -4563,6 +4563,7 @@ int RootClingMain(int argc,
    }
 
    std::ostream &dictStream = (!gOptIgnoreExistingDict && !gOptDictionaryFileName.empty()) ? fileout : std::cout;
+   bool isACLiC = gOptDictionaryFileName.getValue().find("_ACLiC_dict") != std::string::npos;
 
    if (!gOptIgnoreExistingDict) {
       // Now generate a second stream for the split dictionary if it is necessary
@@ -4586,9 +4587,12 @@ int RootClingMain(int argc,
          CreateDictHeader(*splitDictStream, main_dictname);
 
       if (!gOptNoGlobalUsingStd) {
-         AddNamespaceSTDdeclaration(dictStream);
-         if (gOptSplit) {
-            AddNamespaceSTDdeclaration(*splitDictStream);
+         // ACLiC'ed macros might rely on `using namespace std` in front of user headers
+         if (isACLiC) {
+            AddNamespaceSTDdeclaration(dictStream);
+            if (gOptSplit) {
+               AddNamespaceSTDdeclaration(*splitDictStream);
+            }
          }
       }
    }
@@ -4823,6 +4827,15 @@ int RootClingMain(int argc,
          GenerateNecessaryIncludes(dictStream, includeForSource, extraIncludes);
          if (gOptSplit) {
             GenerateNecessaryIncludes(*splitDictStream, includeForSource, extraIncludes);
+         }
+      }
+      if (!gOptNoGlobalUsingStd) {
+         // ACLiC'ed macros might have relied on `using namespace std` in front of user headers
+         if (!isACLiC) {
+            AddNamespaceSTDdeclaration(dictStream);
+            if (gOptSplit) {
+               AddNamespaceSTDdeclaration(*splitDictStream);
+            }
          }
       }
       if (gDriverConfig->fInitializeStreamerInfoROOTFile) {

